pypeit/core/procimg.py

```diff
@@ -9,169 +9,520 @@
+#
+#        # What follows is a special treatment for neighbors, with more relaxed constains.
+#
+#        msgs.info("Finding neighboring pixels affected by cosmic rays")
+#
+#        # We grow these cosmics a first time to determine the immediate neighborhod  :
+#        growcosmics = np.cast['bool'](signal.convolve2d(np.cast['float32'](cosmics), growkernel, mode="same", boundary="symm"))
+#
+#        # From this grown set, we keep those that have sp > sigmalim
+#        # so obviously not requiring sp/f > objlim, otherwise it would be pointless
+#        growcosmics = np.logical_and(sp > sigclip, growcosmics)
+#
+#        # Now we repeat this procedure, but lower the detection limit to sigmalimlow :
+#
+#        finalsel = np.cast['bool'](signal.convolve2d(np.cast['float32'](growcosmics), growkernel, mode="same", boundary="symm"))
+#        finalsel = np.logical_and(sp > sigcliplow, finalsel)
+#
+#        # Unmask saturated pixels:
+#        if satpix is not None:
+#            msgs.info("Masking saturated stars")
+#            finalsel = np.logical_and(np.logical_not(satpix), finalsel)
+#
+#        ncrp = np.sum(finalsel)
+#
+#        msgs.info("{0:5d} pixels detected as cosmics".format(ncrp))
+#
+#        # We find how many cosmics are not yet known :
+#        newmask = np.logical_and(np.logical_not(crmask), finalsel)
+#        nnew = np.sum(newmask)
+#
+#        # We update the mask with the cosmics we have found :
+#        crmask = np.logical_or(crmask, finalsel)
+#
+#        msgs.info("Iteration {0:d} -- {1:d} pixels identified as cosmic rays ({2:d} new)".format(i, ncrp, nnew))
+#        if ncrp == 0:
+#            break
+#
+#    # Additional algorithms (not traditionally implemented by LA cosmic) to
+#    # remove some false positives.
+#    msgs.work("The following algorithm would be better on the rectified, tilts-corrected image")
+#    filt  = ndimage.sobel(sciframe, axis=1, mode='constant')
+#    filty = ndimage.sobel(filt/np.sqrt(np.abs(sciframe)), axis=0, mode='constant')
+#    filty[np.where(np.isnan(filty))]=0.0
+#
+#    sigimg = cr_screen(filty)
+#
+#    sigsmth = ndimage.filters.gaussian_filter(sigimg,1.5)
+#    sigsmth[np.where(np.isnan(sigsmth))]=0.0
+#    sigmask = np.cast['bool'](np.zeros(sciframe.shape))
+#    sigmask[np.where(sigsmth>sigclip)] = True
+#    crmask = np.logical_and(crmask, sigmask)
+#    msgs.info("Growing cosmic ray mask by 1 pixel")
+#    crmask = grow_mask(crmask, grow)
+#
+#    return crmask.astype(bool)
+
+
+# NOTE: This is slower than utils.rebin_evlist by a factor of ~2, but avoids an
+# error when the shape of arr is not an integer multiple of boxcar.
+def boxcar_average(arr, boxcar):
     """
-    Identify cosmic rays using the L.A.Cosmic algorithm
-    U{http://www.astro.yale.edu/dokkum/lacosmic/}
-    (article : U{http://arxiv.org/abs/astro-ph/0108003})
-    This routine is mostly courtesy of Malte Tewes
+    Boxcar average an array.
```

Comment:
Can you provide a bit more information in the docs here. It is unclear to me what this routine operates on. What its the dimensionality of the input array? Also indicate the shapes of array inputs unless your routine is agnostic to that, in which case you should specify. It looks like you arr can have arbitrary shape. If so just say that and indicate how the boxcar smoothing works with these dimensions.

Reply:
The input can be any shape and dimensionality.  I tried to improve the description, but I'm not sure if it helped...